tests/test_flow/test_persistence_fuzz.py

```diff
@@ -96,19 +96,43 @@ def entity_names(
 
         return list(sorted(names))
 
+    def nondeterministic_upstream_entity_names(self, upstream_of):
+        bot_names = interpret.str_or_seq_as_list(upstream_of)
+
+        upstream_names = set()
+        for bot_name in bot_names:
+            bot_entity = self._entities_by_name[bot_name]
+            upstream_names |= bot_entity.all_nondeterministic_upstream_names
+
+        return upstream_names
+
+    def entity_has_nondeterministic_ancestor(self, entity_name):
+        entity = self._entities_by_name[entity_name]
+        return len(entity.all_nondeterministic_upstream_names) > 0
+
+    def entity_is_nondeterministic(self, entity_name):
+        return self._entities_by_name[entity_name].changes_per_run
+
     def called_entity_names(self):
         names = list(self._last_called_names)
         while self._last_called_names:
```

Comment:
I have no recollection of why I didn't just use `list.clear()` here; maybe it was to support Python 2.7?

Reply:
lol not sure, I didn't notice it really. I'll make it use `clear()`.